segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ZKSegmentContainerMonitor.java

```diff
@@ -157,34 +157,40 @@ public void close() {
-        if (desiredList != null) {
-            Collection<Integer> runningContainers = new HashSet<>(this.handles.keySet());
-            Collection<Integer> containersPendingTasks = new HashSet<>(this.pendingTasks);
+            // Fetch the list of containers that is supposed to be owned by this host.
+            Set<Integer> desiredList = getDesiredContainerList();
+            if (desiredList != null) {
+                Collection<Integer> runningContainers = new HashSet<>(this.handles.keySet());
+                Collection<Integer> containersPendingTasks = new HashSet<>(this.pendingTasks);
 
-            // Filter out containers which have pending tasks so we don't initiate conflicting events on the same
-            // containers. Events for these containers will be tried on subsequent runs of this executor.
-            Collection<Integer> containersToBeStarted = CollectionHelpers.filterOut(desiredList, runningContainers);
-            containersToBeStarted = CollectionHelpers.filterOut(containersToBeStarted, containersPendingTasks);
+                // Filter out containers which have pending tasks so we don't initiate conflicting events on the same
+                // containers. Events for these containers will be tried on subsequent runs of this executor.
+                Collection<Integer> containersToBeStarted = CollectionHelpers.filterOut(desiredList, runningContainers);
+                containersToBeStarted = CollectionHelpers.filterOut(containersToBeStarted, containersPendingTasks);
 
-            Collection<Integer> containersToBeStopped = CollectionHelpers.filterOut(runningContainers, desiredList);
-            containersToBeStopped = CollectionHelpers.filterOut(containersToBeStopped, containersPendingTasks);
+                Collection<Integer> containersToBeStopped = CollectionHelpers.filterOut(runningContainers, desiredList);
+                containersToBeStopped = CollectionHelpers.filterOut(containersToBeStopped, containersPendingTasks);
 
-            log.info("Container Changes: Desired = {}, Current = {}, PendingTasks = {}, ToStart = {}, ToStop = {}.",
-                    desiredList, runningContainers, containersPendingTasks, containersToBeStarted,
-                    containersToBeStopped);
+                log.info("Container Changes: Desired = {}, Current = {}, PendingTasks = {}, ToStart = {}, ToStop = {}.",
+                        desiredList, runningContainers, containersPendingTasks, containersToBeStarted,
+                        containersToBeStopped);
 
-            // Initiate the start and stop tasks asynchronously.
-            containersToBeStarted.forEach(this::startContainer);
-            containersToBeStopped.forEach(this::stopContainer);
-        } else {
-            log.warn("No segment container assignments found");
+                // Initiate the start and stop tasks asynchronously.
+                containersToBeStarted.forEach(this::startContainer);
```

Comment:
if we start containers asynchronously and in parallel a subsequent run of this method could start conflicting opearations on these containers (like we could end up with both start and stop container for the same container. should we have some protection against it?